### PR TITLE
Create general purpose Table component

### PR DESCRIPTION
**Pull request by Gkrumbach07**
fixes: https://github.com/opendatahub-io/odh-dashboard/issues/961

## Description

Create a Table component that handles
- structure
- sorting
- pagination

Other toolbar items can be added through a toolbar param

## How Has This Been Tested?

make sure all instances of tables work the same the they did before:
- ~odh-dashboard/frontend/src/pages/BYONImages/BYONImagesTable.tsx~
- ~odh-dashboard/frontend/src/pages/BYONImages/ImportImageModal.tsx~
- ~odh-dashboard/frontend/src/pages/BYONImages/UpdateImageModal.tsx~
- odh-dashboard/frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx
- odh-dashboard/frontend/src/pages/modelServing/screens/projects/ServingRuntimeTable.tsx
- odh-dashboard/frontend/src/pages/modelServing/screens/projects/ServingRuntimeTokensTable.tsx
- odh-dashboard/frontend/src/pages/notebookController/screens/admin/NotebookAdminControl.tsx
- odh-dashboard/frontend/src/pages/projects/screens/detail/data-connections/DataConnectionsTable.tsx
- odh-dashboard/frontend/src/pages/projects/screens/detail/notebooks/NotebookTable.tsx
- odh-dashboard/frontend/src/pages/projects/screens/detail/storage/StorageTable.tsx
- odh-dashboard/frontend/src/pages/projects/screens/projects/ProjectTable.tsx

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: frontend/src/pages/projects/screens/projects/ProjectListView.tsx [hunk 1]
```diff
@@ -42,73 +41,87 @@ const ProjectListView: React.FC<ProjectListViewProps> = ({
...
-              }}
-            />
-          </ToolbarItem>
-          <ToolbarItem>
-            <NewProjectButton />
-          </ToolbarItem>
-          <ToolbarItem>
-            <Button variant="link">
-              <Link to="/notebookController">Launch Jupyter</Link>
+    <React.Fragment>
```
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/projects/ProjectListView.tsx:51]:
```suggestion
    <>
```

File: frontend/src/pages/projects/screens/projects/ProjectListView.tsx [hunk 2]
```diff
@@ -42,73 +41,87 @@ const ProjectListView: React.FC<ProjectListViewProps> = ({
...
+          <ProjectTableRow
+            key={project.metadata.uid}
+            obj={project}
+            isRefreshing={refreshIds.includes(project.metadata.uid || '')}
+            setEditData={(data) => setEditData(data)}
+            setDeleteData={(data) => setDeleteData(data)}
+          />
+        )}
+        toolbarContent={
+          <React.Fragment>
```
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/projects/ProjectListView.tsx:75]:
```suggestion
          <>
```

File: frontend/src/components/Table.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  ToolbarItemVariant,
+} from '@patternfly/react-core';
+import { TableComposable, Thead, Tr, Th, Tbody, Td } from '@patternfly/react-table';
+import React, { useEffect } from 'react';
+import useTableColumnSort, { SortableData } from 'utilities/useTableColumnSort';
+
+type TableProps<DataType> = {
+  data: DataType[];
+  columns: SortableData<DataType>[];
+  Row: (data: DataType) => React.ReactNode;
```
Comment (andrewballantyne) [inline frontend/src/components/Table.tsx:15]: You fixed the issue I was thinking about -- hooks not being able to be fired.

```suggestion
  rowRenderer: (data: DataType) => React.ReactNode;
```

File: frontend/src/pages/projects/screens/projects/ProjectListView.tsx (hunk 2 above)
Comment (Gkrumbach07) [reply frontend/src/pages/projects/screens/projects/ProjectListView.tsx:75]: any reason against react.fragment?

Comment (andrewballantyne) [reply frontend/src/pages/projects/screens/projects/ProjectListView.tsx:78]: Just looks odd to me -- only reason to ever use it is when you need to `key` it as well. I suppose it's okay -- no real issue -- I've done React for a long time and we always linted this out -- so unless we do, I suppose it does not matter.

File: frontend/src/pages/BYONImages/BYONImagesTable.tsx [hunk 4]
```diff
@@ -292,130 +273,113 @@ export const BYONImagesTable: React.FC<BYONImagesTableProps> = ({ images, forceU
...
+        emptyTableView={
           <Tr>
-            <Th />
-            <Th sort={getSortParams(0)}>{columnNames.name}</Th>
-            <Th sort={getSortParams(1)}>{columnNames.description}</Th>
-            <Th>Enable</Th>
-            <Th sort={getSortParams(4)}>{columnNames.user}</Th>
-            <Th sort={getSortParams(5)}>{columnNames.uploaded}</Th>
-            <Th />
+            <Td colSpan={8}>
```
Comment (andrewballantyne) [inline frontend/src/pages/BYONImages/BYONImagesTable.tsx:285]: I wonder if we can infer the Tr and Td colspan code based on columns.length. That way we remove this boilerplate code from everyone.

File: frontend/src/pages/projects/screens/projects/ProjectTable.tsx [hunk 5]
```diff
@@ -1,64 +1,49 @@
...
   refreshProjects: () => Promise<void>;
 };
 
-const ProjectTable: React.FC<ProjectTableProps> = ({
-  clearFilters,
-  projects,
-  getColumnSort,
-  refreshProjects,
-}) => {
+const ProjectTable: React.FC<ProjectTableProps> = ({ clearFilters, projects, refreshProjects }) => {
```
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/projects/ProjectTable.tsx:17]: This is deleted, no?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Looking great! Couple comments -- will have to test tomorrow.

File: frontend/src/pages/BYONImages/BYONImagesTable.tsx (hunk 4 above)
Comment (Gkrumbach07) [reply frontend/src/pages/BYONImages/BYONImagesTable.tsx:285]: It usually is (as i assume this is for empty states). However some empty states have additional logic or are formatted differntly so they are pulled out of the component. 

```tsx
<Tr>
              <Td colSpan={mappedColumns.length} style={{ textAlign: 'center' }}>
                No projects match your filters.{' '}
                <Button variant="link" isInline onClick={clearFilters}>
                  Clear filters
                </Button>
              </Td>
            </Tr>
```

Like this old implementation uses the columns length. colSpan is only ever used in deprecated components (ProjectTable), outdated components (BYON), or inside a sub component. 

we could add a way to add colSpan inside the columns list, but inferring could get confusing to people imo

File: frontend/src/pages/BYONImages/BYONImagesTable.tsx [hunk 6]
```diff
@@ -105,42 +89,39 @@ export const BYONImagesTable: React.FC<BYONImagesTableProps> = ({ images, forceU
...
-  const columnNames = {
-    name: 'Name',
-    description: 'Description',
-    user: 'User',
-    uploaded: 'Uploaded',
-  };
+    {
+      field: 'description',
+      label: 'Description',
+      sortable: (a, b) => (a.description as string).localeCompare(b.description ?? ''),
```
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/pages/BYONImages/BYONImagesTable.tsx:106]: Yep. The BYON Table component is pretty old and does not follow most of the same table design as the other tables. I figure this will be revamped later down the road. 

But yeah i just copied the same logic from the old table, but i can change it to use `??` now

File: frontend/src/components/Table.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        <Tbody>
+          <>
+            {emptyTableView && data.length === 0 && (
+              <Tr>
+                <Td colSpan={columns.length} style={{ textAlign: 'center' }}>
+                  {emptyTableView}
+                </Td>
+              </Tr>
+            )}
+            {data.map((row, i) => rowRenderer(row, i))}
```
Comment (DaoDaoNoCode) [inline frontend/src/components/Table.tsx:97]: You might want to slice the data here based on the pagination settings, or the pagination won't work.
```suggestion
            {data
              .slice(pageSize * (page - 1), pageSize * page)
              .map((row, i) => rowRenderer(row, i))}
```

**Review by DaoDaoNoCode — CHANGES_REQUESTED** [review]
The layout seems broken for some tables... Besides, left 2 comments about pagination.
Screenshot 2023-03-01 at 3 07 51 PM

File: frontend/src/components/Table.tsx (hunk 7 above)
Comment (Gkrumbach07) [reply frontend/src/components/Table.tsx:97]: good catch

File: frontend/src/pages/projects/screens/projects/ProjectListView.tsx [hunk 8]
```diff
@@ -44,72 +44,88 @@ const ProjectListView: React.FC<ProjectListViewProps> = ({
...
-          </ToolbarItem>
-          <ToolbarItem>
-            <NewProjectButton />
-          </ToolbarItem>
-          <ToolbarItem>
-            <Button variant="link">
-              <Link to="/notebookController">Launch Jupyter</Link>
+      <Table
+        enablePagination
+        minPageSize={1}
```
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/pages/projects/screens/projects/ProjectListView.tsx:57]: no, forgot the zero for 10

**Comment by Gkrumbach07** [comment]
> The layout seems broken for some tables

this is b/c some table rows use a tbody when they should really just implement a row

I will go remove the instances of those i find

File: frontend/src/pages/modelServing/screens/projects/ServingRuntimeTableRow.tsx [hunk 9]
```diff
@@ -51,7 +51,7 @@ const ServingRuntimeTableRow: React.FC<ServingRuntimeTableRowProps> = ({
       : undefined;
 
   return (
-    <Tbody isExpanded={isRowExpanded}>
+    <>
       <Tr>
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/projects/ServingRuntimeTableRow.tsx:55]: Um -- this looks like a loss of functionality. @lucferbux 

Based on the PF docs -- https://www.patternfly.org/v4/components/table#tbody -- it looks like `Tbody` should always be true if there is an expand row 🤔 `Tr` is the one who needs `isExpanded` when things are expanded.

File: frontend/src/pages/projects/screens/detail/notebooks/NotebookTableRow.tsx [hunk 10]
```diff
@@ -43,7 +43,7 @@ const NotebookTableRow: React.FC<NotebookTableRowProps> = ({
   const [notebookImage, loaded] = useNotebookImage(obj.notebook);
 
   return (
-    <Tbody isExpanded={isExpanded}>
+    <>
       <Tr>
```
Comment (andrewballantyne) [inline frontend/src/pages/projects/screens/detail/notebooks/NotebookTableRow.tsx:47]: Looks to be the case here too.

File: frontend/src/components/Table.tsx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+                key={col.field + i}
+                sort={col.sortable ? sort.getColumnSort(i) : undefined}
+                width={col.width}
+              >
+                {col.label}
+              </Th>
+            ))}
+          </Tr>
+        </Thead>
+        <Tbody>
```
Comment (andrewballantyne) [inline frontend/src/components/Table.tsx:88]: I wonder if we can set this to `isExpanded` based on a field name of `expand` in the columns.

We'd need to update `frontend/src/utilities/useTableColumnSort.ts` type too. 
Something like `field: 'expand' | string;` to indicate it's a constant value or whatever you want.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Looking good on the replacement -- a couple things are still off on the generic nature of the table.

File: frontend/src/components/Table.tsx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+          </ToolbarItem>
+        </ToolbarContent>
+      </Toolbar>
+      <TableComposable {...props}>
+        {caption && <Caption>{caption}</Caption>}
+        <Thead>
+          <Tr>
+            {columns.map((col, i) => (
+              <Th
+                key={col.field + i}
```
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/components/Table.tsx:79]: Some tables used the index

[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/src/components/Table.tsx:79]: Ah, okay -- don't think it's needed here. Use unique headers. This is likely a problem with your empty `field` states you have in all the various columns. `field: ''` is probably invalid... but TS doesn't have length tests for us to use.

[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/components/Table.tsx:79]: Its all the BYON notebook tables

[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/src/components/Table.tsx:79]: Maybe you revert this effort then -- if you can't really properly support that table, we can do it later. Let us not compromise the base table implementation because of old code.

File: frontend/src/components/Table.tsx [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+          <>
+            {emptyTableView && data.length === 0 && (
+              <Tr>
+                <Td colSpan={columns.length} style={{ textAlign: 'center' }}>
+                  {emptyTableView}
+                </Td>
+              </Tr>
+            )}
+            {data
+              .slice(pageSize * (page - 1), pageSize * page)
```
[the remark this answers is not in the document]
Comment (Gkrumbach07) [reply frontend/src/components/Table.tsx:98]: Is useMemo allowable here?

```tsx
  const data = React.useMemo(() => enablePagination ? allData.slice(pageSize * (page - 1), pageSize * page) : allData, [allData, enablePagination, page, pageSize])
```

File: frontend/src/pages/modelServing/screens/projects/ServingRuntimeTableRow.tsx (hunk 9 above)
Comment (Gkrumbach07) [reply frontend/src/pages/modelServing/screens/projects/ServingRuntimeTableRow.tsx:55]: Right. these row components will have two rows with one always being visible and one have isExpanded. There is no reason to have it on a tbody

```tsx
<Tr isExpanded={isRowExpanded}>
        <ServingRuntimeTableExpandedSection
          activeColumn={expandedColumn}
          obj={obj}
          onClose={() => onExpandColumn(undefined)}
          onDeployModel={onDeployModal}
        />
</Tr>
```

File: frontend/src/pages/projects/screens/detail/notebooks/NotebookTableRow.tsx (hunk 10 above)
Comment (Gkrumbach07) [reply frontend/src/pages/projects/screens/detail/notebooks/NotebookTableRow.tsx:47]: I dont think this breaks functionality?

File: frontend/src/components/Table.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/src/components/Table.tsx:98]: You can -- not sure you need to though. We don't deal with 1000s of data points. We'll have other issues before that.

File: .env [hunk 14]
```diff
@@ -3,7 +3,7 @@
 #OC_USER=specify_in_.env
 #OC_PASSWORD=specify_in_.env
 #OC_TOKEN=specify_in_.env
-#OC_PROJECT=specify_in_.env
+OC_PROJECT=redhat-ods-applications
```
Comment (lucferbux) [inline .env:6]: You should not modify the `.env` file, this is a placeholder.

Please revert this change and if you don't want this to be tracked, create a new `.env.local` file and add there the override!

**Review by lucferbux — CHANGES_REQUESTED** [review]

File: frontend/src/components/Table.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/src/components/Table.tsx:81]: We should swing back in here and remove the index when we can make sure the field is not an empty string -- might go into the `useTableSortColumn` function to throw an error.

**Review by andrewballantyne — COMMENTED** [review]
Testing now...

File: frontend/src/components/Table.tsx [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+      <Toolbar>
+        <ToolbarContent>
+          {toolbarContent}
+          {showPagination && (
+            <ToolbarItem variant="pagination" alignment={{ default: 'alignRight' }}>
+              {pagination('down')}
+            </ToolbarItem>
+          )}
+        </ToolbarContent>
+      </Toolbar>
```
Comment (andrewballantyne) [inline frontend/src/components/Table.tsx:74]: Toolbar needs to not be rendered if you have no toolbarContent and you're not showing pagination.

You created header space:
image

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: .env (hunk 14 above)
Comment (Gkrumbach07) [reply .env:6]: dang i left it in

File: frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx [hunk 16]
```diff
@@ -106,7 +106,6 @@ const InferenceServiceListView: React.FC<InferenceServiceListViewProps> = ({
         clearFilters={resetFilters}
         servingRuntimes={servingRuntimes}
         inferenceServices={filteredInferenceServices.slice(pageSize * (page - 1), pageSize * page)}
```
Comment (DaoDaoNoCode) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx:108]: If you are going to do the pagination in the table, I think you don't need to do the slice here, right? you can just simply pass all the `filteredInferenceService` into `InferenceServiceTable`.

File: frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx [hunk 17]
```diff
@@ -78,46 +58,33 @@ const InferenceServiceListView: React.FC<InferenceServiceListViewProps> = ({
...
-        getColumnSort={sortInferenceService.getColumnSort}
+        inferenceServices={filteredInferenceServices}
         refresh={refresh}
-      />
-      {showPagination && (
-        <Toolbar>
-          <ToolbarContent>
-            <ToolbarItem variant="pagination" alignment={{ default: 'alignRight' }}>
-              {pagination('up')}
+        enablePagination={true}
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx:64]: Nit.
```suggestion
        enablePagination
```

File: frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx [hunk 18]
```diff
@@ -26,8 +25,6 @@ const InferenceServiceListView: React.FC<InferenceServiceListViewProps> = ({
   } = React.useContext(ModelServingContext);
   const [searchType, setSearchType] = React.useState<SearchType>(SearchType.NAME);
   const [search, setSearch] = React.useState('');
-  const [page, setPage] = React.useState(1);
-  const [pageSize, setPageSize] = React.useState(MIN_PAGE_SIZE);
   const sortInferenceService = useTableColumnSort<InferenceServiceKind>(
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceListView.tsx:27]: I think you want to remove this too -- no? `.transformData` is the sort transform -- which is done by your Table.

File: frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx [hunk 19]
```diff
@@ -1,28 +1,30 @@
...
 import DeleteInferenceServiceModal from './DeleteInferenceServiceModal';
 
 type InferenceServiceTableProps = {
   clearFilters?: () => void;
   inferenceServices: InferenceServiceKind[];
   servingRuntimes: ServingRuntimeKind[];
-  getColumnSort: GetColumnSort;
   refresh: () => void;
+  enablePagination?: boolean;
+  toolbarContent?: React.ReactElement<typeof ToolbarItem>;
```
Comment (andrewballantyne) [inline frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx:18]: Interesting, any reason you didn't take my "infer type from Table prop"?

Curious if it looks bulky or if you had a specific objection to it.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx (hunk 19 above)
Comment (Gkrumbach07) [reply frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx:18]: do you mean like this or merging the types
```tsx
  enablePagination?: TableProps<InferenceServiceKind>["enablePagination"];
  toolbarContent?:TableProps<InferenceServiceKind>["toolbarContent"]
ts

Comment (andrewballantyne) [reply frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx:18]: Content props (that take JSX) are often better to infer... same goes for functions that have complex signatures. Booleans are up to you -- but if you're threading props through, it doesn't hurt to just infer. Saves the extra step if we ever change the prop later.

You could also use `Pick` (antithesis of `Omit`) and just say "these are being threaded" https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys

You may want to wrap it in `Partial` if it doesn't do it for you 🤔 https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype

Comment (Gkrumbach07) [reply frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx]: Ok, I used pick, as it seemed the most "future proof"

Comment (andrewballantyne) [reply frontend/src/pages/modelServing/screens/global/InferenceServiceTable.tsx]: I created a playground to test -- see it [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwCwAUIzAJ5hZwCyLACsWOgC8cAN6M4EuKhJYAKgAtgAOwDmALjgAjCBAA2OJfQaSpMgMKLdAEzMQlMLPY3Y8MAHQv8AOQhWsRgF9GRgB6ELgAdXlkeGACOABJOAB3ZHsUKytgGGA7ZF04MH50GGj4BLJSACsAV1R4UspkLNU4RogalXkAfkZcO3q4WR1dTWQoBIdSYQAKAEo4QQA+UThKGBqoJTglGt1dGjgAw7DTEA5UdlxgfLa2Dgh4-vA7R3hkjlT7YIZWdjhbC8lG8EgBVVBYPgQASLUTiSTSc6lZQqBJWOwwLxYLBWDS7ECaLBQIwmdEKFEabR6AwkyTQnJ5XS2exvbrOKjuTwwACi+nO9gAPH8sI8hiMxhMpktAj9hVwWABBfA1fJQmHCMTGBEycmqSkjGnwiSIrAWYDWZkOJyYDkeDk+PyBOAAMjgPGAuAA1gLAZBgfYwRC1agADTkMmKVRkOAAH3I9NySnylreZGlP36SkGIEVytV-HByBUWA03CVG3z0KEcK1xp1kfUbSgNSwIaNZ1NlhsGLeeL2ujbtbgEYpTZbYdOSQAIgB5HbYqzDiC6lRwAC0a7g2RSaDWWAAjjVgJQrO3J0ulGR4MCcXAE4yU+kN1v3rv70ndIxjj9TlEYlv4iSL4GnkLBSBgCAtA4fICnfW4CGgOAczgOosFwNAsF6JhfnueVyxVXQADF+xYGcwAZD9uQADxgKBkGDWFNRME0V31ak0lpOtzjNC0e2tLk7VcB1-C-F03XGHJ8gFd0vR9Eg-RBcFIWKMMyBHKNY3jcjE2TPiYDTdNsMzbNcwrXQ1ULYsACZS1MgjiP2UjtMZajaPo4pGPbFiGw0Wjxy88wu0fGA+32QcTFOdTGz81s4HPdFL2vBclxXddN23VJ0Dgz8h3iuwr3nW9suCtKXx3LLnI-L8aCAA).

**Review by andrewballantyne — APPROVED** [review]

**Review by andrewballantyne — APPROVED** [review]
